assembly/src/compiled_library.rs

```diff
@@ -0,0 +1,184 @@
+pub struct ProcedureInfo {
+    pub name: ProcedureName,
+    pub digest: RpoDigest,
+}
+
+/// Represents a library where all modules modules were compiled into a [`MastForest`].
+pub struct CompiledLibrary {
+    mast_forest: MastForest,
+    // a path for every `root` in the associated MAST forest
+    exports: Vec<CompiledFullyQualifiedProcedureName>,
+    metadata: CompiledLibraryMetadata,
+}
+
+/// Constructors
+impl CompiledLibrary {
+    /// Constructs a new [`CompiledLibrary`].
+    pub fn new(
+        mast_forest: MastForest,
+        exports: Vec<CompiledFullyQualifiedProcedureName>,
+        metadata: CompiledLibraryMetadata,
+    ) -> Result<Self, CompiledLibraryError> {
+        if mast_forest.procedure_roots().len() != exports.len() {
+            return Err(CompiledLibraryError::InvalidExports {
+                exports_len: exports.len(),
+                roots_len: mast_forest.procedure_roots().len(),
+            });
+        }
+
+        Ok(Self {
+            mast_forest,
+            exports,
+            metadata,
+        })
+    }
+}
+
+impl CompiledLibrary {
+    /// Returns the inner [`MastForest`].
+    pub fn mast_forest(&self) -> &MastForest {
+        &self.mast_forest
+    }
+
+    /// Returns the fully qualified name of all procedures exported by the library.
+    pub fn exports(&self) -> &[CompiledFullyQualifiedProcedureName] {
+        &self.exports
+    }
+
+    /// Returns the library metadata.
+    pub fn metadata(&self) -> &CompiledLibraryMetadata {
+        &self.metadata
+    }
+
+    /// Returns an iterator over the module infos of the library.
+    pub fn into_module_infos(self) -> impl Iterator<Item = ModuleInfo> {
+        let mut modules_by_path: BTreeMap<LibraryPath, ModuleInfo> = BTreeMap::new();
+
+        for (proc_index, proc_name) in self.exports.into_iter().enumerate() {
+            modules_by_path
+                .entry(proc_name.module_path.clone())
+                .and_modify(|compiled_module| {
+                    let proc_node_id = self.mast_forest.procedure_roots()[proc_index];
+                    let proc_digest = self.mast_forest[proc_node_id].digest();
+
+                    compiled_module.add_procedure_info(ProcedureInfo {
+                        name: proc_name.name.clone(),
+                        digest: proc_digest,
+                    })
+                })
+                .or_insert_with(|| {
+                    let proc_node_id = self.mast_forest.procedure_roots()[proc_index];
+                    let proc_digest = self.mast_forest[proc_node_id].digest();
+                    let proc = ProcedureInfo {
+                        name: proc_name.name,
+                        digest: proc_digest,
+                    };
+
+                    ModuleInfo::new(proc_name.module_path, vec![proc])
+                });
+        }
+
+        modules_by_path.into_values()
+    }
+}
+
+pub struct CompiledLibraryMetadata {
+    pub name: LibraryNamespace,
```

Comment:
We don't want to use `LibraryNamespace` here - that's actually a detail of Miden Assembly's namespacing semantics that no longer applies to libraries (i.e. `LibraryNamespace` is now a misnomer).

More importantly though, MAST libraries will contain code from modules that may or may not belong to the same MASM namespace, so the name associated with a `Library` is basically giving a name to all of those items as a group. While we can re-use `LibraryNamespace` for this, I think that might convey the idea that all of the contents are in the same MASM namespace like the previous iteration of libraries (MASL).

I'd probably just use `Ident` here.

Reply:
This was removed for now